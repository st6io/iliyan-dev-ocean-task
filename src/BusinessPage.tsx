import { x } from '@xstyled/styled-components';

const BusinessPage = () => <x.div bg="skyblue">Business Page</x.div>;

export default BusinessPage;
