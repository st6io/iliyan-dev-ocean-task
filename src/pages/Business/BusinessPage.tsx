import { useParams } from 'react-router-dom';

import { Props, x } from '@xstyled/styled-components';

import Error from '@components/Error';
import Layout from '@components/Layout';
import LoadingPlaceholder from '@components/LoadingPlaceholder';
import Section from '@components/Section';
import Table, { LoadingPlaceholder as TablePlaceholder } from '@components/Table';

import nearbyPlaces from './nearby-places.mock.json';

import { useBusinessQuery } from '../hooks';

const nearbyPlacesRows = nearbyPlaces.map(({ name, address }) => ({
  id: name,
  cells: [name, address],
}));

const SectionGroup = (props: Props) => <x.div display="flex" py={8} px={7} {...props} />;

const BusinessPage = () => {
  const { businessId } = useParams();
  const { loading, error, data } = useBusinessQuery(businessId);

  const { name, image, address, phone, email } = data?.business || {};
  const { number, street, city, zip, country } = address || {};

  if (error) {
    return (
      <Layout>
        <Error />
      </Layout>
    );
  }

  return (
    <Layout>
      <x.div h="40vh">
        {loading ? (
          <LoadingPlaceholder />
        ) : (
          <x.img src={image} alt={`${name} cover photo`} w={1} h="100%" objectFit="cover"></x.img>
        )}
      </x.div>

      <x.div display="flex" justifyContent="space-between" my={8} mx={12}>
        <SectionGroup justifyContent="space-around" flex={4} mr={5}>
          {loading ? (
            <LoadingPlaceholder height="50%" />
          ) : (
            <>
              <Section title="address" mr={10}>
                <Section.Text>{`${number} ${street}`}</Section.Text>

                <Section.Text>{`${city}, ${country} ${zip}`}</Section.Text>
              </Section>

              <Section title="contact">
                <Section.Text>{phone}</Section.Text>

                <Section.Text>{email}</Section.Text>
              </Section>
            </>
          )}
        </SectionGroup>

        <SectionGroup bg="white" flex={5} flexDirection="column">
          {loading ? (
            <TablePlaceholder rowsCount={10} />
          ) : (
            <Section title="nearby places" flex={1}>
              <Table variant="secondary" rows={nearbyPlacesRows} w={1} />
            </Section>
          )}
        </SectionGroup>
      </x.div>
    </Layout>
  );
};

export default BusinessPage;
