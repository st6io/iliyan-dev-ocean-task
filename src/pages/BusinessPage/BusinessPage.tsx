import { Props, x } from '@xstyled/styled-components';

import Layout from '@components/Layout';
import Table from '@components/Table';

import Section from './Section';
import nearbyPlaces from './nearby-places.mock.json';

import businessesData from '../businesses.mock.json';

const { name, image, address, phone, email } = businessesData[0];
const { number, street, city, zip, country } = address;

const nearbyPlacesRows = nearbyPlaces.map(({ name, address }) => ({
  id: name,
  cells: [name, address],
}));

const SectionGroup = (props: Props) => <x.div display="flex" py={8} px={7} {...props} />;

const BusinessPage = () => (
  <Layout>
    <x.img src={image} alt={`${name} cover photo`} w={1} h="40vh" objectFit="cover"></x.img>

    <x.div display="flex" justifyContent="space-between" my={8} mx={12}>
      <SectionGroup justifyContent="space-around" flex={4} mr={5}>
        <Section title="address" mr={10}>
          <Section.Text>{`${number} ${street}`}</Section.Text>

          <Section.Text>{`${city}, ${country} ${zip}`}</Section.Text>
        </Section>

        <Section title="contact">
          <Section.Text>{phone}</Section.Text>

          <Section.Text>{email}</Section.Text>
        </Section>
      </SectionGroup>

      <SectionGroup bg="white" flex={5}>
        <Section title="nearby places" flex={1}>
          <Table variant="secondary" rows={nearbyPlacesRows} w={1} />
        </Section>
      </SectionGroup>
    </x.div>
  </Layout>
);

export default BusinessPage;
