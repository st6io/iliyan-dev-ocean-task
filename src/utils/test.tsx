import { ReactElement, ReactNode } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { render as nativeRender } from '@testing-library/react';

const render = (ui: ReactElement, { path = '/', ...options } = {}) => {
  const RouterWrapper = ({ children }: { children: ReactNode }) => (
    <MemoryRouter initialEntries={[path]}>{children}</MemoryRouter>
  );

  return nativeRender(ui, {
    wrapper: RouterWrapper,
    ...options,
  });
};

export * from '@testing-library/react';
export { render };
