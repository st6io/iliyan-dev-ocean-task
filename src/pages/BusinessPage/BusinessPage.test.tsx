import { render, screen } from '@src/utils/test';

import BusinessPage from './BusinessPage';
import nearbyPlaces from './nearby-places.mock.json';

const nearbyPlacesRows = nearbyPlaces.map(({ name, address }, index) => ({
  id: `name`,
  cells: [name, address],
}));

const renderPage = () => render(<BusinessPage />);

describe('BusinessPage', () => {
  it('should match snapshot', () => {
    const { container } = renderPage();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render image cover with correct styles', () => {
    renderPage();

    const [, coverImg] = screen.queryAllByRole('img');

    expect(coverImg).toBeInTheDocument();
    expect(coverImg).toHaveAttribute(
      'src',
      'https://images.unsplash.com/photo-1527015175922-36a306cf0e20?crop=entropy&cs=tinysrgb&fit=crop&fm=jpg&h=600&w=1000',
    );
    expect(coverImg).toHaveAttribute('alt', 'Zazio cover photo');
    expect(coverImg).toHaveStyle({
      width: '100%',
      height: '40vh',
      'object-fit': 'cover',
    });
  });

  it('should apply correct spacing between section groups', () => {
    renderPage();

    const [, img] = screen.queryAllByRole('img');
    const contentElement = img.nextSibling;

    expect(contentElement).toHaveStyle({
      display: 'flex',
      'justify-content': 'space-between',
    });
  });

  describe('Address', () => {
    it('should apply correct styles to section "Address"', () => {
      renderPage();

      const addressTitle = screen.queryByText('address');
      expect(addressTitle).toBeInTheDocument();
      expect(addressTitle).toHaveStyle({
        'font-size': '1.5rem',
        color: 'rgb(53, 52, 73)',
        'margin-bottom': '2rem',
        'text-transform': 'capitalize',
      });

      const addressSection = addressTitle?.parentNode;
      expect(addressSection).toHaveStyle({
        'margin-right': '2.5rem',
      });
    });

    it('should render full address', () => {
      renderPage();

      expect(screen.getByText('84586 Straubel')).toBeInTheDocument();
      expect(screen.getByText('Manchester, United Kingdom SG4')).toBeInTheDocument();
    });
  });

  describe('Contact', () => {
    it('should apply correct styles to section "Contact"', () => {
      renderPage();

      const contactTitle = screen.queryByText('contact');
      expect(contactTitle).toBeInTheDocument();
      expect(contactTitle).toHaveStyle({
        'font-size': '1.5rem',
        color: 'rgb(53, 52, 73)',
        'margin-bottom': '2rem',
        'text-transform': 'capitalize',
      });
    });
  });

  describe('Nearby Places', () => {
    it('should render section group with background', () => {
      renderPage();

      const placesTitle = screen.queryByText('nearby places');
      const sectionGroup = placesTitle?.parentNode?.parentNode;
      expect(sectionGroup).toHaveStyle({
        'background-color': 'rgb(255, 255, 255)',
      });
    });

    it('should use table with "secondary" variant', () => {
      renderPage();

      const trElements = screen.queryByRole('table')?.querySelectorAll('tr') || [];
      expect(trElements.length).toBeGreaterThan(0);
      trElements.forEach((rowElement) => {
        expect(rowElement).toHaveStyle({
          'background-color': 'rgb(248, 248, 250)',
        });
      });

      nearbyPlacesRows
        .flatMap(({ cells }) => cells)
        .forEach((cell) => {
          const cellElement = screen.queryByText(cell);
          expect(cellElement).toBeInTheDocument();
          expect(cellElement).toHaveStyle({
            color: 'rgb(146, 146, 157)',
          });
        });
    });
  });
});
