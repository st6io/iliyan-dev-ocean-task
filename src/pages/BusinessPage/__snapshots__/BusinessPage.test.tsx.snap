// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessPage should match snapshot 1`] = `
<div
  class="sc-cxabCf FKOKy"
>
  <header
    class="sc-efBctP eZxVPd"
  >
    <a
      href="/"
    >
      <img
        alt="home logo"
        class="sc-BeQoi"
        src="logo.png"
      />
    </a>
  </header>
  <div
    class="sc-cxabCf bsgFzP"
  >
    <img
      alt="Zazio cover photo"
      class="sc-BeQoi lpbvHV"
      src="https://images.unsplash.com/photo-1527015175922-36a306cf0e20?crop=entropy&cs=tinysrgb&fit=crop&fm=jpg&h=600&w=1000"
    />
    <div
      class="sc-cxabCf hinPhb"
    >
      <div
        class="sc-cxabCf fBBkYE"
      >
        <section
          class="sc-bWXABl kZOqy"
        >
          <div
            class="sc-cxabCf sc-afnQL  eEWMdl"
          >
            address
          </div>
          <div
            class="sc-cxabCf sc-kngDgl  jNZoFN"
          >
            84586 Straubel
          </div>
          <div
            class="sc-cxabCf sc-kngDgl  jNZoFN"
          >
            Manchester, United Kingdom SG4
          </div>
        </section>
        <section
          class="sc-bWXABl"
        >
          <div
            class="sc-cxabCf sc-afnQL  eEWMdl"
          >
            contact
          </div>
          <div
            class="sc-cxabCf sc-kngDgl  jNZoFN"
          >
            609-306-3610
          </div>
          <div
            class="sc-cxabCf sc-kngDgl  jNZoFN"
          >
            igowanlock0@networkadvertising.org
          </div>
        </section>
      </div>
      <div
        class="sc-cxabCf bNCOOk"
      >
        <section
          class="sc-bWXABl kgzIGE"
        >
          <div
            class="sc-cxabCf sc-afnQL  eEWMdl"
          >
            nearby places
          </div>
          <table
            class="sc-WCkqM sc-grVGCS GQgxF eKfFBe"
          >
            <tbody
              class="sc-ikjQzJ"
            >
              <tr
                class="sc-jGprRt bzOoij"
              >
                <td
                  class="sc-tsFYE sc-iJkHyd fMNVnW JjpSb"
                >
                  Caff√® Nero
                </td>
                <td
                  class="sc-tsFYE sc-iJkHyd fMNVnW JjpSb"
                >
                  1st Floor, c/o Morleys, 472-488 Brixton Rd, London SW9 8EH
                </td>
              </tr>
              <tr
                class="sc-jGprRt bzOoij"
              >
                <td
                  class="sc-tsFYE sc-iJkHyd fMNVnW JjpSb"
                >
                  EAT.
                </td>
                <td
                  class="sc-tsFYE sc-iJkHyd fMNVnW JjpSb"
                >
                  2nd Floor, c/o Morleys, 472-488 Brixton Rd, London SW9 8EH
                </td>
              </tr>
              <tr
                class="sc-jGprRt bzOoij"
              >
                <td
                  class="sc-tsFYE sc-iJkHyd fMNVnW JjpSb"
                >
                  Pret A Manager
                </td>
                <td
                  class="sc-tsFYE sc-iJkHyd fMNVnW JjpSb"
                >
                  3rd Floor, c/o Morleys, 472-488 Brixton Rd, London SW9 8EH
                </td>
              </tr>
              <tr
                class="sc-jGprRt bzOoij"
              >
                <td
                  class="sc-tsFYE sc-iJkHyd fMNVnW JjpSb"
                >
                  Coffee Republic
                </td>
                <td
                  class="sc-tsFYE sc-iJkHyd fMNVnW JjpSb"
                >
                  4th Floor, c/o Morleys, 472-488 Brixton Rd, London SW9 8EH
                </td>
              </tr>
            </tbody>
          </table>
        </section>
      </div>
    </div>
  </div>
</div>
`;
